### DISCUSSION TITLE
Pallet-Connectors-Gateway

### DISCUSSION THREAD

**Discussion by mustermeiszer** [centrifuge/centrifuge-chain#1358] [General]
Discussion ground for the gateway until we have a PR.

**Comment by mustermeiszer** [comment]
cc: @NunoAlexandre @cdamian @wischli

**Comment by mustermeiszer** [comment]
Notes for the following state of the code: https://github.com/centrifuge/centrifuge-chain/commit/44227eb150373627483d6cb7b980d5d5e0183d04

1. https://github.com/centrifuge/centrifuge-chain/blob/44227eb150373627483d6cb7b980d5d5e0183d04/pallets/connectors-gateway/connectors-gateway-routers/src/axelar_evm.rs#L37-L39

Yes, we need that int the router, as we need a way to know what to call. 

2. https://github.com/centrifuge/centrifuge-chain/blob/44227eb150373627483d6cb7b980d5d5e0183d04/pallets/connectors-gateway/connectors-gateway-routers/src/axelar_evm.rs#L79-L84

Not sure if we need something different TBH. I think we should use the same derivation that we are using normally for native to `H160`. Branan can help you out in case. Here is the [spec](https://centrifuge.hackmd.io/nEwvthS8TLOPDAMZ191-rA?both) for this

3. https://github.com/centrifuge/centrifuge-chain/blob/44227eb150373627483d6cb7b980d5d5e0183d04/pallets/connectors-gateway/connectors-gateway-routers/src/axelar_evm.rs#L91-L92

I think we must look into the frontier implementation of `trait Runner` for this. I would guess, we need a nonce...

4. https://github.com/centrifuge/centrifuge-chain/blob/44227eb150373627483d6cb7b980d5d5e0183d04/pallets/connectors-gateway/connectors-gateway-routers/src/axelar_evm.rs#L93

Not sure. Need to look at frontier again.

**Comment by mustermeiszer** [comment]
~~I think we do not need the `struct EVMChain` as the domain, hence chain can be determined by the storage of the router itself. WDYT?~~

Nevermind, I saw you need that for the encoding of the Axelar message. I would add a doc comment for the struct then that this is needed for Axelar routing or so. But really nice work already!!!

**Comment by mustermeiszer** [comment]
@cdamian In the production networks, calling pallet-evm will be only allowed for root. So we should change:

```rust
                pallet_evm::Pallet::<T>::call(
                        // The `call` method uses `EnsureAddressTruncated` to check this `origin`, this ensures
                        // that `origin` is the same as `source`.
                        RawOrigin::Signed(sender.clone()).into(),
```

to

```rust
                pallet_ethereum::Pallet::<T>::transact(
                        // The `call` method uses `EnsureAddressTruncated` to check this `origin`, this ensures
                        // that `origin` is the same as `source`.
                        pallet_ethereum::RawOrigin::EthereumTransaction(sender).into(),
```

The sender should be converted with our standard converter. See `runtime::common`

**Reply by cdamian** [reply]
Got it! I did not get to test this yet but I will make this change and try it out soon.

**Comment by cdamian** [comment]
Opened a draft PR, let's move any other discussions here - https://github.com/centrifuge/centrifuge-chain/pull/1376/files
